### DISCUSSION TITLE
About MemoryBarrier

### DISCUSSION THREAD

**Discussion by KMwin** [winsiderss/systeminformer#2092] [General Discussion]
**What is the purpose of MemoryBarrier in this context:**

_IRQL_requires_max_(APC_LEVEL)
LONG KphGetDriverUnloadProtectionCount(
    VOID
    )
{
    LONG count;

    PAGED_CODE();

    count = KphpDriverUnloadProtectionRef.Count;
    MemoryBarrier();

    return count;
}

**Comment by dmex** [comment]
Some processors execute code out-of-order and with an incorrect sequence. MemoryBarrier guarantees the order of execution as written and returns the correct value for count which is required in this case because the function is used for memory management and an incorrect value would leak memory - Most compilers insert fences automatically during compile time but sometimes an explicit reference is needed to guarantee the correct placement.

See the examples section of the wiki:
https://en.wikipedia.org/wiki/Memory_barrier

**Reply by KMwin** [reply]
"Thank you for your response. I have another question: What are the differences between 

count = KphpDriverUnloadProtectionRef.Count;
MemoryBarrier();
return count;

 and 

MemoryBarrier();
count = KphpDriverUnloadProtectionRef.Count;
return count;

**Reply by dmex** [reply]
The same equivalent of changing:
 
fwrite() 
flush() 

vs

flush() aka MemoryBarrier();
fwrite() aka count = KphpDriverUnloadProtectionRef.Count

--------------
ULONG count; // local cpu instruction store buffer

count = KphpDriverUnloadProtectionRef.Count; // Value could be in global memory, CPU instruction/load/store cache buffers
MemoryBarrier(); // Wait for instruction/store buffer to reach zero or wait for flush before continuing

**Reply by KMwin** [reply]
Thanks a lot, the answer is perfectly clear!

**Comment by dmex** [comment]
The project has Github Copilot and for fun I asked it to explain the function:

![image](https://github.com/winsiderss/systeminformer/assets/1306177/788cf66a-e16c-434c-b812-c3984d161f8f)
![image](https://github.com/winsiderss/systeminformer/assets/1306177/a0f8ce47-2812-4132-b7b4-be0c1820477b)

**Reply by KMwin** [reply]
If this function(KphGetDriverUnloadProtectionCount) is not inlined by the compiler,Is it possible for the following situation?

1. The core A has completed MemoryBarrier() and count = KphpDriverUnloadProtectionRef.Count.
2. Core B modifies the value of KphpDriverUnloadProtectionRef.Count.
3. The RET instruction is executed at this point in Core A.
